File: vscodeExt/219.contains-duplicate-ii.py
#
# @lc app=leetcode id=219 lang=python3
#
# [219] Contains Duplicate II
#

# @lc code=start
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        s = set()
        # l = 0 
        # r = k  # 初始窗口为[0, k]
        i = 0 
        while i < len(nums):
            if nums[i] in s:
                return True
            else:
                s.add(nums[i])

            if len(s) == k+1:
                s.remove(nums[i-k])


            i+=1    
        return False
        
# @lc code=end

File: vscodeExt/219.contains-duplicate-ii.java
import java.util.HashSet;
import java.util.Set;

/*
 * @lc app=leetcode id=219 lang=java
 *
 * [219] Contains Duplicate II
 */

// @lc code=start
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        boolean res = false;
        Set<Integer> s = new HashSet<>();
        for(int i =0;i<=k && i<nums.length;i++){
            if(s.contains(nums[i])){
                return true;
            }
            s.add(nums[i]);
        }
        for(int i=k+1; i<nums.length;i++){
            s.remove(nums[i-k-1]);
            if(s.contains(nums[i])){
                res = true;
                break;
            } else {
                s.add(nums[i]);
            }
        }

        return res;
        
    }
}
// @lc code=end

